File: macos/UniviewCodeBuddyProxy/Sources/Services/StreamTranslator.swift
import Foundation
import Hummingbird

// ═══════════════════════════════════════════════════════════════
// StreamTranslator — OpenAI → OpenAI SSE 流透传翻译器
// 解析上游 SSE chunk，替换 model/id，清理 choices，输出标准 OpenAI 格式
// ═══════════════════════════════════════════════════════════════

final class StreamTranslator {

    // MARK: - State

    /// 首个有效 token 时间
    private var tfft: Date?
    /// prompt token 计数
    private var promptTokens: Int = 0
    /// completion token 计数
    private var completionTokens: Int = 0
    /// reasoning token 计数（来自 completion_tokens_details）
    private var reasoningTokens: Int = 0

    /// 客户端请求的模型名，用于替换上游返回的 model 字段
    private let requestedModel: String
    /// 替换后的请求 ID
    private let requestID: String

    // MARK: - Init

    init(requestedModel: String) {
        self.requestedModel = requestedModel
        self.requestID = "chatcmpl-" + Self.randomHex(12)
    }

    // MARK: - Process

    /// 处理一行上游 SSE 数据，返回零条或多条要写入客户端的 SSE 字符串
    func process(line: String) -> [String] {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return [] }

        let (dataStr, done, ok) = Self.parseSSELine(trimmed)
        guard ok else { return [] }

        // 记录 TFFT
        if tfft == nil {
            tfft = Date()
        }

        if done {
            // 流结束 — 可在此报告遥测
            if reasoningTokens > 0 {
                NSLog("stream: model=\(requestedModel) prompt_tokens=\(promptTokens) completion_tokens=\(completionTokens) reasoning_tokens=\(reasoningTokens)")
            }
            return ["data: [DONE]\n\n"]
        }

        guard let data = dataStr.data(using: .utf8),
              var chunk = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return []
        }

        // 提取 usage
        if let usage = chunk["usage"] as? [String: Any] {
            if let pt = usage["prompt_tokens"] as? Int { promptTokens = pt }
            else if let pt = usage["prompt_tokens"] as? Double { promptTokens = Int(pt) }

            if let ct = usage["completion_tokens"] as? Int { completionTokens = ct }
            else if let ct = usage["completion_tokens"] as? Double { completionTokens = Int(ct) }

            if let details = usage["completion_tokens_details"] as? [String: Any] {
                if let rt = details["reasoning_tokens"] as? Int { reasoningTokens = rt }
                else if let rt = details["reasoning_tokens"] as? Double { reasoningTokens = Int(rt) }
            }
        }

        // 替换 model 和 id
        if chunk["choices"] != nil {
            chunk["model"] = requestedModel
            chunk["id"] = requestID
            Self.cleanChunkChoices(&chunk)
        }

        guard let cleaned = try? JSONSerialization.data(withJSONObject: chunk),
              let cleanedStr = String(data: cleaned, encoding: .utf8) else {
            return []
        }

        return ["data: \(cleanedStr)\n\n"]
    }

    // MARK: - Error Output

    /// 生成 SSE 错误输出（OpenAI 格式）
    static func sseError(message: String) -> String {
        let msg = message
            .replacingOccurrences(of: "\"", with: "\\\"")
            .replacingOccurrences(of: "\n", with: "\\n")
        return "data: {\"error\":{\"message\":\"\(msg)\",\"type\":\"upstream_error\"}}\n\ndata: [DONE]\n\n"
    }

    /// 生成上下文超限 SSE 错误（OpenAI 格式）
    static func sseContextLimitError(message: String) -> String {
        let msg = "request too large: " + message
            .replacingOccurrences(of: "\"", with: "\\\"")
            .replacingOccurrences(of: "\n", with: "\\n")
        return "data: {\"error\":{\"message\":\"\(msg)\",\"type\":\"invalid_request_error\"}}\n\ndata: [DONE]\n\n"
    }

    // MARK: - Collect Chunks (Non-Streaming)

    /// 从上游字节流收集所有 chunk，返回 CollectedResult（用于非流式请求）
    static func collectChunks(from bytes: URLSession.AsyncBytes, requestedModel: String) async throws -> CollectedResult {
        var result = CollectedResult(model: requestedModel)
        let requestID = "chatcmpl-" + randomHex(12)
        result.id = requestID

        for try await line in bytes.lines {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard !trimmed.isEmpty else { continue }

            let (dataStr, done, ok) = parseSSELine(trimmed)
            guard ok, !done else { continue }

            guard let data = dataStr.data(using: .utf8),
                  let chunk = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                continue
            }

            // 提取 choices
            if let choices = chunk["choices"] as? [[String: Any]] {
                for choice in choices {
                    if let delta = choice["delta"] as? [String: Any] {
                        if let content = delta["content"] as? String, !content.isEmpty {
                            result.content += content
                        }
                        if let reasoning = delta["reasoning_content"] as? String, !reasoning.isEmpty {
                            result.reasoningContent += reasoning
                        }
                        if let toolCalls = delta["tool_calls"] as? [[String: Any]] {
                            for tcMap in toolCalls {
                                let idx = intFromMap(tcMap, key: "index")

                                // 确保 toolCalls 数组够长
                                while result.toolCalls.count <= idx {
                                    result.toolCalls.append([
                                        "id": "",
                                        "type": "function",
                                        "function": ["name": "", "arguments": ""] as [String: Any],
                                    ] as [String: Any])
                                }

                                if let id = tcMap["id"] as? String, !id.isEmpty {
                                    if var tc = result.toolCalls[idx] as? [String: Any] {
                                        tc["id"] = id
                                        result.toolCalls[idx] = tc
                                    }
                                }
                                if let fn = tcMap["function"] as? [String: Any] {
                                    if var tc = result.toolCalls[idx] as? [String: Any],
                                       var existing = tc["function"] as? [String: Any] {
                                        if let name = fn["name"] as? String, !name.isEmpty {
                                            existing["name"] = name
                                        }
                                        if let args = fn["arguments"] as? String, !args.isEmpty {
                                            let prevArgs = existing["arguments"] as? String ?? ""
                                            existing["arguments"] = prevArgs + args
                                        }
                                        tc["function"] = existing
                                        result.toolCalls[idx] = tc
                                    }
                                }
                            }
                        }
                    }
                    if let fr = choice["finish_reason"] as? String, !fr.isEmpty {
                        result.finishReason = fr
                    }
                }
            }

            // 提取 usage
            if let usage = chunk["usage"] as? [String: Any] {
                if let pt = usage["prompt_tokens"] as? Int { result.promptTokens = pt }
                else if let pt = usage["prompt_tokens"] as? Double { result.promptTokens = Int(pt) }

                if let ct = usage["completion_tokens"] as? Int { result.completionTokens = ct }
                else if let ct = usage["completion_tokens"] as? Double { result.completionTokens = Int(ct) }

                if let details = usage["prompt_tokens_details"] as? [String: Any] {
                    if let ct = details["cached_tokens"] as? Int, ct > 0 { result.cacheCreationTokens = ct }
                    else if let ct = details["cached_tokens"] as? Double, Int(ct) > 0 { result.cacheCreationTokens = Int(ct) }
                    if let cct = details["cache_creation_tokens"] as? Int, cct > 0 { result.cacheCreationTokens = cct }
                    else if let cct = details["cache_creation_tokens"] as? Double, Int(cct) > 0 { result.cacheCreationTokens = Int(cct) }
                }
                if let details = usage["completion_tokens_details"] as? [String: Any] {
                    if let rt = details["reasoning_tokens"] as? Int { result.reasoningTokens = rt }
                    else if let rt = details["reasoning_tokens"] as? Double { result.reasoningTokens = Int(rt) }
                }
            }
        }

        return result
    }

    // MARK: - Private Helpers

    /// 解析 SSE 行，提取 data 内容
    private static func parseSSELine(_ line: String) -> (data: String, done: Bool, ok: Bool) {
        var dataStr = line
        if line.hasPrefix("data: ") {
            dataStr = String(line.dropFirst(6))
        } else if line.hasPrefix("data:") {
            dataStr = String(line.dropFirst(5))
        } else {
            return ("", false, false)
        }

        dataStr = dataStr.trimmingCharacters(in: .whitespaces)

        if dataStr == "[DONE]" {
            return ("", true, true)
        }

        if dataStr.isEmpty {
            return ("", false, false)
        }

        return (dataStr, false, true)
    }

    /// 清理 chunk choices，只保留标准字段（原地修改）
    private static func cleanChunkChoices(_ chunk: inout [String: Any]) {
        guard var choices = chunk["choices"] as? [[String: Any]] else { return }

        for i in 0..<choices.count {
            var choice = choices[i]

            if var delta = choice["delta"] as? [String: Any] {
                // 只保留 role, content, tool_calls, reasoning_content
                let allowedDeltaKeys: Set<String> = ["role", "content", "tool_calls", "reasoning_content"]
                for key in delta.keys where !allowedDeltaKeys.contains(key) {
                    delta.removeValue(forKey: key)
                }
                // 移除空的 tool_calls 数组
                if let tcs = delta["tool_calls"] as? [[String: Any]], tcs.isEmpty {
                    delta.removeValue(forKey: "tool_calls")
                }
                choice["delta"] = delta
            }

            // 只保留 index, delta, finish_reason
            let allowedChoiceKeys: Set<String> = ["index", "delta", "finish_reason"]
            for key in choice.keys where !allowedChoiceKeys.contains(key) {
                choice.removeValue(forKey: key)
            }

            // 规范化 finish_reason
            if let fr = choice["finish_reason"] as? String, !fr.isEmpty {
                switch fr {
                case "stop", "tool_calls", "length", "content_filter":
                    break
                default:
                    choice["finish_reason"] = "stop"
                }
            }

            choices[i] = choice
        }

        chunk["choices"] = choices
    }

    /// 从字典中提取整数值（兼容 Int 和 Double）
    private static func intFromMap(_ map: [String: Any], key: String) -> Int {
        if let v = map[key] as? Int { return v }
        if let v = map[key] as? Double { return Int(v) }
        return 0
    }

    /// 生成随机十六进制字符串
    private static func randomHex(_ n: Int) -> String {
        var bytes = [UInt8](repeating: 0, count: n)
        _ = SecRandomCopyBytes(kSecRandomDefault, n, &bytes)
        return bytes.map { String(format: "%02x", $0) }.joined()
    }
}

File: windows/UniviewCodeBuddyProxy.Server/Services/StreamTranslator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;

namespace UniviewCodeBuddyProxy.Services;

/// <summary>
/// OpenAI to OpenAI SSE passthrough translator.
/// Replaces model/id, cleans chunk choices, extracts usage.
/// </summary>
public sealed class StreamTranslator
{
    // ═══ State ═══

    private int _promptTokens;
    private int _completionTokens;
    private int _reasoningTokens;

    private readonly string _requestedModel;
    private readonly string _requestID;

    // ═══ Constructor ═══

    public StreamTranslator(string requestedModel)
    {
        _requestedModel = requestedModel;
        _requestID = "chatcmpl-" + RandomHex(12);
    }

    // ═══ Process ═══

    /// <summary>
    /// Process a single upstream SSE line, returning zero or more output SSE strings.
    /// </summary>
    public List<string> ProcessLine(string line)
    {
        var trimmed = line.Trim();
        if (string.IsNullOrEmpty(trimmed)) return [];

        var (dataStr, done, ok) = ParseSSELine(trimmed);
        if (!ok) return [];

        if (done)
            return ["data: [DONE]\n\n"];

        JsonElement chunk;
        try
        {
            chunk = JsonDocument.Parse(dataStr).RootElement;
        }
        catch
        {
            return [];
        }

        // Extract usage
        ExtractUsage(chunk);

        // Replace model and id if there are choices
        string outputJson;
        if (chunk.TryGetProperty("choices", out _))
        {
            var modified = ModifyChunk(chunk);
            outputJson = modified.GetRawText();
        }
        else
        {
            outputJson = chunk.GetRawText();
        }

        return [$"data: {outputJson}\n\n"];
    }

    // ═══ Error Output ═══

    /// <summary>Generate SSE error output (OpenAI format).</summary>
    public static string SseError(string message)
    {
        var msg = EscapeForSSE(message);
        return $"data: {{\"error\":{{\"message\":\"{msg}\",\"type\":\"upstream_error\"}}}}\n\ndata: [DONE]\n\n";
    }

    /// <summary>Generate context limit SSE error (OpenAI format).</summary>
    public static string SseContextLimitError(string message)
    {
        var msg = EscapeForSSE("request too large: " + message);
        return $"data: {{\"error\":{{\"message\":\"{msg}\",\"type\":\"invalid_request_error\"}}}}\n\ndata: [DONE]\n\n";
    }

    // ═══ Collect Chunks (Non-Streaming) ═══

    /// <summary>
    /// Collect all upstream chunks and return a CollectedResult.
    /// </summary>
    public static CollectedResult CollectChunks(System.IO.Stream stream, string requestedModel)
    {
        var result = new CollectedResult { Model = requestedModel };
        var requestID = "chatcmpl-" + RandomHex(12);
        result.ID = requestID;

        using var reader = new System.IO.StreamReader(stream);
        string? line;
        while ((line = reader.ReadLine()) != null)
        {
            var trimmed = line.Trim();
            if (string.IsNullOrEmpty(trimmed)) continue;

            var (dataStr, done, ok) = ParseSSELine(trimmed);
            if (!ok || done) continue;

            JsonElement chunk;
            try
            {
                chunk = JsonDocument.Parse(dataStr).RootElement;
            }
            catch
            {
                continue;
            }

            // Extract choices
            if (chunk.TryGetProperty("choices", out var choices) && choices.ValueKind == JsonValueKind.Array)
            {
                foreach (var choice in choices.EnumerateArray())
                {
                    if (choice.TryGetProperty("delta", out var delta))
                    {
                        if (delta.TryGetProperty("content", out var contentEl))
                        {
                            var content = contentEl.GetString();
                            if (!string.IsNullOrEmpty(content))
                                result.Content += content;
                        }
                        if (delta.TryGetProperty("reasoning_content", out var reasoningEl))
                        {
                            var reasoning = reasoningEl.GetString();
                            if (!string.IsNullOrEmpty(reasoning))
                                result.ReasoningContent += reasoning;
                        }
                        if (delta.TryGetProperty("tool_calls", out var toolCallsEl) && toolCallsEl.ValueKind == JsonValueKind.Array)
                        {
                            foreach (var tcMap in toolCallsEl.EnumerateArray())
                            {
                                var idx = IntFromMap(tcMap, "index");

                                // Ensure toolCalls array is long enough
                                while (result.ToolCalls.Count <= idx)
                                {
                                    result.ToolCalls.Add(new Dictionary<string, object>
                                    {
                                        ["id"] = "",
                                        ["type"] = "function",
                                        ["function"] = new Dictionary<string, object>
                                        {
                                            ["name"] = "",
                                            ["arguments"] = "",
                                        },
                                    });
                                }

                                if (tcMap.TryGetProperty("id", out var idEl))
                                {
                                    var id = idEl.GetString();
                                    if (!string.IsNullOrEmpty(id))
                                        result.ToolCalls[idx]["id"] = id;
                                }
                                if (tcMap.TryGetProperty("function", out var fnEl))
                                {
                                    var tcDict = result.ToolCalls[idx];
                                    var fnDict = (Dictionary<string, object>)tcDict["function"];

                                    if (fnEl.TryGetProperty("name", out var nameEl))
                                    {
                                        var name = nameEl.GetString();
                                        if (!string.IsNullOrEmpty(name))
                                            fnDict["name"] = name;
                                    }
                                    if (fnEl.TryGetProperty("arguments", out var argsEl))
                                    {
                                        var args = argsEl.GetString();
                                        if (!string.IsNullOrEmpty(args))
                                            fnDict["arguments"] = (fnDict.GetValueOrDefault("arguments", "")?.ToString() ?? "") + args;
                                    }
                                }
                            }
                        }
                    }
                    if (choice.TryGetProperty("finish_reason", out var frEl))
                    {
                        var fr = frEl.GetString();
                        if (!string.IsNullOrEmpty(fr))
                            result.FinishReason = fr;
                    }
                }
            }

            // Extract usage
            if (chunk.TryGetProperty("usage", out var usage))
            {
                if (usage.TryGetProperty("prompt_tokens", out var pt))
                    result.PromptTokens = GetInt(pt);
                if (usage.TryGetProperty("completion_tokens", out var ct))
                    result.CompletionTokens = GetInt(ct);
                if (usage.TryGetProperty("prompt_tokens_details", out var ptd))
                {
                    if (ptd.TryGetProperty("cached_tokens", out var cachedEl))
                    {
                        var val = GetInt(cachedEl);
                        if (val > 0) result.CacheCreationTokens = val;
                    }
                    if (ptd.TryGetProperty("cache_creation_tokens", out var cctEl))
                    {
                        var val = GetInt(cctEl);
                        if (val > 0) result.CacheCreationTokens = val;
                    }
                }
                if (usage.TryGetProperty("completion_tokens_details", out var ctd))
                {
                    if (ctd.TryGetProperty("reasoning_tokens", out var rtEl))
                        result.ReasoningTokens = GetInt(rtEl);
                }
            }
        }

        return result;
    }

    // ═══ Private Helpers ═══

    private void ExtractUsage(JsonElement chunk)
    {
        if (!chunk.TryGetProperty("usage", out var usage)) return;

        if (usage.TryGetProperty("prompt_tokens", out var pt))
            _promptTokens = GetInt(pt);
        if (usage.TryGetProperty("completion_tokens", out var ct))
            _completionTokens = GetInt(ct);
        if (usage.TryGetProperty("completion_tokens_details", out var details) &&
            details.TryGetProperty("reasoning_tokens", out var rt))
            _reasoningTokens = GetInt(rt);
    }

    private JsonElement ModifyChunk(JsonElement chunk)
    {
        // We need to deserialize, modify, and re-serialize
        var dict = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(chunk.GetRawText()) ?? new();
        dict["model"] = JsonSerializer.Deserialize<JsonElement>($"\"{_requestedModel}\"");
        dict["id"] = JsonSerializer.Deserialize<JsonElement>($"\"{_requestID}\"");
        CleanChunkChoices(dict);
        var json = JsonSerializer.Serialize(dict);
        return JsonDocument.Parse(json).RootElement;
    }

    private static void CleanChunkChoices(Dictionary<string, JsonElement> chunk)
    {
        if (!chunk.TryGetValue("choices", out var choicesEl)) return;
        var choices = JsonSerializer.Deserialize<List<Dictionary<string, JsonElement>>>(choicesEl.GetRawText());
        if (choices == null) return;

        var allowedDeltaKeys = new HashSet<string> { "role", "content", "tool_calls", "reasoning_content" };
        var allowedChoiceKeys = new HashSet<string> { "index", "delta", "finish_reason" };
        var validFinishReasons = new HashSet<string> { "stop", "tool_calls", "length", "content_filter" };

        for (var i = 0; i < choices.Count; i++)
        {
            var choice = choices[i];

            // Clean delta
            if (choice.TryGetValue("delta", out var deltaEl))
            {
                var deltaDict = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(deltaEl.GetRawText()) ?? new();
                var deltaKeysToRemove = deltaDict.Keys.Where(k => !allowedDeltaKeys.Contains(k)).ToList();
                foreach (var key in deltaKeysToRemove)
                    deltaDict.Remove(key);

                // Remove empty tool_calls arrays
                if (deltaDict.TryGetValue("tool_calls", out var tcsEl) && tcsEl.ValueKind == JsonValueKind.Array && tcsEl.GetArrayLength() == 0)
                    deltaDict.Remove("tool_calls");

                choice["delta"] = JsonSerializer.Deserialize<JsonElement>(JsonSerializer.Serialize(deltaDict));
            }

            // Remove non-standard choice keys
            var choiceKeysToRemove = choice.Keys.Where(k => !allowedChoiceKeys.Contains(k)).ToList();
            foreach (var key in choiceKeysToRemove)
                choice.Remove(key);

            // Normalize finish_reason
            if (choice.TryGetValue("finish_reason", out var frEl) && frEl.ValueKind == JsonValueKind.String)
            {
                var fr = frEl.GetString() ?? "";
                if (!string.IsNullOrEmpty(fr) && !validFinishReasons.Contains(fr))
                {
                    choice["finish_reason"] = JsonSerializer.Deserialize<JsonElement>("\"stop\"");
                }
            }

            choices[i] = choice;
        }

        chunk["choices"] = JsonSerializer.Deserialize<JsonElement>(JsonSerializer.Serialize(choices));
    }

    private static (string Data, bool Done, bool Ok) ParseSSELine(string line)
    {
        var dataStr = line;
        if (line.StartsWith("data: "))
            dataStr = line[6..];
        else if (line.StartsWith("data:"))
            dataStr = line[5..];
        else
            return ("", false, false);

        dataStr = dataStr.Trim();

        if (dataStr == "[DONE]")
            return ("", true, true);

        if (string.IsNullOrEmpty(dataStr))
            return ("", false, false);

        return (dataStr, false, true);
    }

    private static int IntFromMap(JsonElement map, string key)
    {
        if (!map.TryGetProperty(key, out var el)) return 0;
        if (el.ValueKind == JsonValueKind.Number) return el.GetInt32();
        return 0;
    }

    private static int GetInt(JsonElement el)
    {
        if (el.ValueKind == JsonValueKind.Number) return el.GetInt32();
        return 0;
    }

    private static string RandomHex(int n)
    {
        var bytes = new byte[n];
        Random.Shared.NextBytes(bytes);
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }

    private static string EscapeForSSE(string s)
    {
        return s.Replace("\\", "\\\\").Replace("\"", "\\\"").Replace("\n", "\\n");
    }
}

/// <summary>Collected result from non-streaming upstream chunks.</summary>
public sealed class CollectedResult
{
    public string ID { get; set; } = "";
    public string Model { get; set; } = "";
    public string Content { get; set; } = "";
    public string ReasoningContent { get; set; } = "";
    public List<Dictionary<string, object>> ToolCalls { get; set; } = [];
    public string FinishReason { get; set; } = "";
    public int PromptTokens { get; set; }
    public int CompletionTokens { get; set; }
    public int ReasoningTokens { get; set; }
    public int CacheCreationTokens { get; set; }
}
